### PR TITLE
APIs modularization

### PR DESCRIPTION
**Pull request by miguelangel-dev**
## Description
In this PR, we have reviewed the nef-modularization to simplify how users should integrate nef in their projects.

- Users just need to `import nef` to use nef in their projects.
- Upgrade Swift-Package-Manager. You can find two `Package` files, one of them in `/` to build the `nef` library; and the second one in `/project` to build the binaries.
- Add support to `FP` (added dependency to Bow) - functional apis to Carbon, Jekyll, Markdown

## APIs
You can find more information in the [file](https://github.com/bow-swift/nef/blob/api_modules/project/Component/nef/API.swift)

### `@_exported`
`modulemap` is only available for C libs, so we can not create a custom layout of how symbols visibility will be through our umbrella-library. Apple has an undocumented kind of import `@_exported` that let us expose symbols of nested libraries. You can read more about it in the [following thread](https://forums.swift.org/t/exported-and-fixing-import-visibility/9415)

On this way, when the user `import net` automatically is importing `NefModels` (we expose the symbols of NefModels from nef library, using @_exported)

## Example of use
I have migrated `nef-plugin` to this API version, using the new APIs, you can review it [here](https://github.com/bow-swift/nef-plugin/pull/10)

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -2,8 +2,12 @@
 **/.DS_Store
 **/xcuserdata/
 BowPlayground
-/.build
+
+## SPM
+.build
+Package.resolved
```
Comment (truizlop) [inline .gitignore:8]: Apparently, `Package.resolved` should not be ignored to ensure reproducibility of builds.

File: project/Component/NefModels/CarbonModel.swift [hunk 2]
```diff
@@ -121,16 +121,16 @@ extension CarbonStyle.Color {
 public struct CarbonError: Error {
     public let filename: String
     public let snippet: String
-    public let error: CarbonErrorOption
+    public let error: CarbonError.Option
     
-    public init(filename: String, snippet: String, error: CarbonErrorOption) {
+    public init(filename: String, snippet: String, error: CarbonError.Option) {
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]: Suggestion: Rename `Option` to `Cause`, I think it is more meaningful. Likewise, the `error` field should be renamed to `cause`.

File: project/Component/nef/API.swift [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+import BowEffects
+
+
+public enum Render: RenderAPI, RenderFP { case build
+    public enum Page: PageAPI, PageFP {   case build
+    }
+}
+
+
+public protocol RenderAPI {
```
Comment (truizlop) [inline project/Component/nef/API.swift]: I don't think we should be providing a non-functional API, and even less doing this distinction in naming. I'd just provide an API that is based on Bow Effects. Moreover, the non-functional API can always be derived from the functional one, so whoever wants it, can easily build it themselves.

**Review by truizlop — CHANGES_REQUESTED** [review]

File: project/Component/NefModels/CarbonModel.swift (hunk 2 above)
Comment (miguelangel-dev) [reply project/Component/NefModels/CarbonModel.swift]: I like it :)

File: .gitignore (hunk 1 above)
Comment (miguelangel-dev) [reply .gitignore:8]: It is a different case, in this case, this `Package.resolved` is generated after build nef library locally for testing purposes; it is not the dependencies for the project - we are using Xcode SPM.

Comment (truizlop) [reply .gitignore:8]: 👍

File: project/Component/NefModels/CarbonModel.swift [hunk 4]
```diff
@@ -121,16 +121,16 @@ extension CarbonStyle.Color {
 public struct CarbonError: Error {
     public let filename: String
     public let snippet: String
-    public let error: CarbonErrorOption
+    public let error: CarbonError.Cause
```
Comment (truizlop) [inline project/Component/NefModels/CarbonModel.swift]: Replace: `error` -> `cause`

File: project/Component/nef/API.swift [hunk 5]
```diff
@@ -0,0 +1,61 @@
+//  Copyright © 2019 The nef Authors.
+
+import Foundation
+@_exported import NefModels
```
Comment (truizlop) [inline project/Component/nef/API.swift:4]: Add a comment in the PR description describing the `@_exported` and a link to where it is documented, for future reference.

File: project/Component/nef/CarbonAPI.swift [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                            },
+                            failure: { error in
+                                callback(.left(.invalidSnapshot))
+                            })
+            }^
+        }
+        
+        guard !Thread.isMainThread else {
+            fatalError("carbonIO(_ carbon:,output:) should be invoked in background thread")
+        }
```
Comment (truizlop) [inline project/Component/nef/CarbonAPI.swift:31]: Is this really necessary? If we are switching to the main queue ourselves, the precondition shouldn't be necessary. Also, the message of the `fatalError` is outdated.

**Review by truizlop — CHANGES_REQUESTED** [review]

**Review by truizlop — APPROVED** [review]

**Pull request merged**